src/Compilers/Core/AnalyzerDriver/AnalyzerExecutor.cs

```diff
@@ -466,7 +476,18 @@ internal static bool CanHaveExecutableCodeBlock(ISymbol symbol)
 
         internal void ExecuteAndCatchIfThrows(DiagnosticAnalyzer analyzer, Action analyze)
         {
-            ExecuteAndCatchIfThrows(analyzer, analyze, _onAnalyzerException, _cancellationToken);
+            object gate;
+            if (_singleThreadedAnalyzerToGateMap.TryGetValue(analyzer, out gate))
+            {
+                lock(gate)
```

Comment:
Async semaphore lies in the workspaces layer, and needs to be moved down to the core shared collections layer. I'll add this work item to #664